### ISSUE TITLE
SDAG closures not opened for 'forall' header

### ISSUE THREAD

**Issue by PhilMiller** [UIUC-PPL/charm#380]
*Original issue: https://charm.cs.illinois.edu/redmine/issues/380*

---
From Joseph Rock:

While working on my project, I noticed a neat little bug with the SDAG `forall` construct.

Let's say we want to execute a 'forall' inside a 'when' clause:

```
when entryMethodA(int totalIterations) {
  forall [i] (1 : totalIterations, 1) {
    //do work here
  }
}
```

The variable 'totalIterations' should be in scope for the 'forall' construct. But when we compile, the .def.h code gets a little bit wrong. It'll throw an error like "totalIterations" not in scope.

The offending line looks like: `int __first = (1), __last = (totalIterations), __stride = (1);`
This is the 'forall''s iteration construct. It appears that there is direct textual substitution from the SDAG code to the generated .def.h code.

Yet, the closures (such as entryMethodA) are passed to this method as parameters. Thus, the 'forall' generated code has access to the parameter called totalIterations. Instead of "totalIterations", though, it would be something like "gen0->getP0()"; (gen0 is the closure object representing the "when entryMethodA" scope)

Were 'forall's intended to have dynamic iteration? Given that it has access to its parent closures' parameters, I think it is in the realm of possibility.

**Comment by PhilMiller** [comment]
*Original date: 2013-12-20 16:53:05*

---
As I noted in email:
> I believe the issue described is that the closure in question should be 'opened' at the point where the forall bounds are being evaluated.
Jonathan agreed.

**Comment by lifflander** [comment]
*Original date: 2013-12-21 01:45:38*

---
bug solved, see commit 0d5460af2a6ac48e241ec940f586bfb8f804c384

**Comment by lifflander** [comment]
*Original date: 2013-12-21 01:50:51*

---
see gerrit issues for bug fix and corresponding test:

http://charm.cs.uiuc.edu/gerrit/#/c/26/
http://charm.cs.uiuc.edu/gerrit/#/c/27/
